options.go

```diff
@@ -9,13 +9,40 @@ import (
 	"time"
 
 	"github.com/Shopify/sarama"
+	"github.com/lovoo/goka/headers"
 	"github.com/lovoo/goka/storage"
 )
 
+// UpdateContext defines the interface for UpdateCallback arguments.
+type UpdateContext interface {
+	// Storage returns the storage that should be updated with the input message.
+	Storage() storage.Storage
+
+	// Topic returns the topic of input message.
+	Topic() Stream
+
+	// Key returns the key of the input message.
+	Key() string
+
+	// Partition returns the partition of the input message.
+	Partition() int32
+
+	// Offset returns the offset of the input message.
+	Offset() int64
+
+	// Value returns the value of the input message.
+	Value() []byte
+
+	// Headers returns the headers of the input message.
+	//
+	// It is recommended to lazily evaluate the headers to reduce overhead per message
+	// when headers are not used.
+	Headers() Headers
+}
+
 // UpdateCallback is invoked upon arrival of a message for a table partition.
-// The partition storage shall be updated in the callback.
-// Headers can converted to a map using FromSarama function.
-type UpdateCallback func(s storage.Storage, partition int32, key string, value []byte, headers ...*sarama.RecordHeader) error
+// The partition storage in the context shall be updated in the callback.
+type UpdateCallback func(ctx UpdateContext) error
```

Comment:
I hate to be the picky/pedantic douche here, but to be real honest, moving all arguments into the single interface looks a bit strange to me. 
The use case of the update callback is in 99% of the cases: execute default i.e. store `key`/`value` in `storage`. 
So I think those three values (+ the partition for backwards compatibility) should be part of the callback function. The rest is sugar on top and only rarely used, if someone needs the headers and whatnot, so it can be wrapped in the object. 
Having all default values hidden behind functions also adds three extra function calls, that cannot be inlined. The performance overhead is probably not measurable as the storage will be much slower, but still it seems like an unnecessary penalty.
In my head, the type would look like this:
`type UpdateCallback func(s storage.Storage, partition int32, key string, value []byte, ctx UpdateContext) error`
(maybe the linter will complain that `ctx` is not the first argument, then we could rename it).
What do you think? 
Still, many thanks for your work here and sorry for the pickyness. Let me know if you need support.

Reply:
Yep, the performance overhead is negligible in the context of IO. This is similar to the processor callback context which is probably executed much more often in the wild, and might not be using persistence which means it's more likely for any overhead to manifest in that path.

I see your point about optimizing for the main usecase, but having both callbacks follow the same pattern is a good argument as well.

Thoughts?

Reply:
Not convinced yet :) 
The performance was actually not my point, shouldn't have brought it up. The main point was that the function's purpose is to store key/value in storage, so those three things should be part of the arguments and not hidden behind some context. 
This would also match the processor-callback scheme, which contains the centric item (the message) directly in the parameters, not in the context: `func(ctx goka.Context, msg interface{})`.

Then the default is simple:
```go
func update(s storage.Strorage, key string, value []byte, ctx updateContext){
    s.Set(key, value)
}
```
Which makes semantically much more sense to me than
```go
func update(ctx updateContext){
    ctx.Storage().Set(ctx.Key(), ctx.Value())
}
```
Don't you think?

I mean we wouldn't write a sum function like this
```go
type ops interface{
 A() float64
 B() float64
 SetResult(float64)
}
func sum(values ops){
    ops.SetResult(ops.A()+ops.B())
}
```
Right?
We'd simply write
```go
func sum(a, b float64)float64{
    return a+b
}
```

I'd also leave the partitionID argument, so basically leave the function as it was before, just replace the `sarama.Headers` with the generic `UpdateContext`, to minimize the breaking changes.
What do you think?

Reply:
Good point Franz. I'm convinced by the centric-item argument :)

Considering we are making a breaking change, Reordering the arguments to `UpdateContext, Storage, Key, Value` and moving partition to `UpdateContext` (since it sounds like it's rarely used) looks more idiomatic go to me. The callback might get `conext.Context` at some point which should be expected to be (within) the first argument.

```go
func update(ctx UpdateContext, s storage.Strorage, key string, value []byte){
    s.Set(key, value)
}
```
Do I have your vote on that one?

Reply:
The reordering and removing partition makes total sense.
Love it :heart:

Reply:
Done.